docs/docs/reference-react-component.md

```diff
@@ -230,34 +230,59 @@ componentWillUnmount()
 ### `setState()`
 
 ```javascript
-setState(nextState, callback)
+setState(updater, [callback])
 ```
 
-Performs a shallow merge of nextState into current state. This is the primary method you use to trigger UI updates from event handlers and server request callbacks.
+`setState()` is an asynchronous method which enqueues changes to be made to the state during the next update cycle. This is the primary method you use to trigger updates in response to event handlers, server responses, etc...
 
-The first argument can be an object (containing zero or more keys to update) or a function (of state and props) that returns an object containing keys to update.
+`setState()` does not immediately mutate `this.state` but creates a pending state transition. Accessing `this.state` after calling this method can potentially return the previous state, rather than the state after enqueued updates have been applied. This is a common source of bugs in React applications.
```

Comment:
This is better, but I'm concerned that it implies that `this.state` is a source of truth. You can mostly get away with this assumption in sync mode, with the exception that you describe here. But this won't be the case with async.

What I'd like to write is

> You should only read `this.state` (or `this.props`) inside a lifecycle method or `render`. It's not safe in an event handler, promise handler, timeout callback, or any place in the stack above React, where you should use `this.setState((state, props) => stateUpdate)` instead.

but this might be too prescriptive. Because again, this isn't really an issue in sync mode, only with async. And we'll likely have to come up with a new API for async mode, anyway, at which point this `setState` doc will be moot.

Maybe this is a good way to split the difference:

> `setState()` does not always immediately update the component. It may batch or defer the update until later. This makes reading `this.state` right after calling `setState()` a potential pitfall. Instead, use `componentDidUpdate` or a `setState` callback (`setState(update, callback)`), either of which are guaranteed to fire after the update has been applied.